File: [0490][Medium][The_Maze]/The_Maze.py
# -*- coding: utf-8 -*-  
# leetcode time     cost : 328 ms
# leetcode memory   cost : 14.8 MB
# Solution 1, DFS
class Solution:
    def hasPath(self, maze: List[List[int]], start: List[int], destination: List[int]) -> bool:
        m,n = len(maze),len(maze[0])
        Visited = [[0]*n for _ in range(m)]
        Directions = [(-1,0,'u'),(0,1,'r'),(0,-1,'l'),(1,0,'d')]
        # set init value
        Visited[start[0]][start[1]] = 0
        return  DFS(maze,start,Visited)  
        def isValid(x,y,maze):
            if 0<=x<m and 0<=y<n and maze[x][y] == 0:
                return True
            else:
                return False
        def destinationReached(x,y,destination):
            if x == destination[0] and y == destination[1]:
                return True
            else:
                return False 

        def DFS(maze,start,Visited):
            pop_x,pop_y = start[0],start[1]
            for dx,dy,direction in Directions:
                x,y = pop_x+dx, pop_y+dy  # 1st step from start is 0
                # if destination is a hole like 499, then while need also check [x,y] != destination
                while isValid(x,y,maze) :                    
                    x,y = x+dx, y+dy
                                       
                # backtrack to the point before the blocked point, try another direction
                x,y = x-dx, y-dy
                if destinationReached(x,y,destination):
                    #print("destinationReached",x,y,",Visited[x][y] ",Visited[x][y],",on the direction:",direction)
                    return True
                
                # update the Visited[x][y] to 1 and drill down if Visited[x][y]==0
                if Visited[x][y]==0:
                    Visited[x][y]=1
                    #if dest found in this drill down, directly return, else continue the for loop
                    if DFS(maze,[x,y],Visited):
                        return DFS(maze,[x,y],Visited)
            return False
              

File: [0490][Medium][The_Maze]/The_Maze.java
// leetcode time     cost : 2 ms
// leetcode memory   cost : 42.9 MB 
// Time  Complexity: O(MN)
// Space Complexity: O(MN)
// solution 1, DFS
public class Solution {
    public boolean hasPath(int[][] maze, int[] start, int[] destination) {
        boolean[][] visited = new boolean[maze.length][maze[0].length];
        visited[start[0]][start[1]] = false;
        return dfs(maze, start, destination, visited);
    }
    public boolean dfs(int[][] maze, int[] start, int[] destination, boolean[][] visited) {
        int[][] dirs={{0, 1}, {0, -1}, {-1, 0}, {1, 0}};
        //since start and dest are not the same, no need check dest at first.

        for (int[] dir: dirs) {
            int x = start[0] + dir[0];
            int y = start[1] + dir[1];
            while (x >= 0 && y >= 0 && x < maze.length && y < maze[0].length && maze[x][y] == 0) {
                x += dir[0];
                y += dir[1];
            }
            // backtrack to the point before the blocked point, try another direction
            x= x-dir[0];y = y-dir[1] ; 
            //check dest found or not
            if (x == destination[0] && y == destination[1])
                return true; 
            //if (x,y) not visited, drilled down                         
            if (!visited[x][y]) {
                visited[x][y] =true;
                int[] newStart = {x,y};
                System.out.println("drill down to x: "+x+" ,y: "+y);
                if (dfs(maze, newStart, destination, visited))
                    return true;
            }
        }
        return false;
    }
}